papermc: provide multiple versions

MayNiklas opened the pull request:
## Description of changes

When hosting a Minecraft server, many players want to stay on a specific version.
Updates often have a big impact, on how the game is being played.
Also: compatibility with mods varies a lot!

My suggestion: adding variants of the papermc packages.
This allows players, to lock a version they want to play on.
Without this, a flake update would upgrade the Minecraft server to a new version, that would break gameplay in a lot of cases.

Since the `papermc` points to the latest version, this PR only adds a easy way to specify a papermc version.

I also verified, that overrides still work!
```nix
  my-paper = papermc.overrideAttrs (finalAttrs: previousAttrs:
    let mcVersion = "1.20.2"; buildNum = "243"; in
    {
      version = "${mcVersion}.${buildNum}";
      src = pkgs.fetchurl {
        url = "https://papermc.io/api/v2/projects/paper/versions/${mcVersion}/builds/${buildNum}/downloads/paper-${mcVersion}-${buildNum}.jar";
        hash = "sha256-rJSgspLZz6LVGfeOuadCtL9Y4PKgrturfzOFzfxs540=";
      };
    });
```
-> works as it did before my change!

Additionally, this now also works:
```nix
  my-paper = papermc.overrideAttrs {
    version = "1.20.2.242";
    hash = "sha256-rJSgspLZz6LVGfeOuydCtL9Y4PKgrturfzOFzfxs540=";
  };
```

I also added myself to meta.maintainers.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nonchris commented:
this tastes like sugar and we need it for navigating our maze of sharks!  
thanks for your work - very useful :)

aaronjanse commented on pkgs/games/papermc/default.nix:
```diff
@@ -1,50 +1,69 @@
+          buildNum = builtins.elemAt (lib.splitVersion version) 3;
+        in
+        fetchurl {
+          url = "https://papermc.io/api/v2/projects/paper/versions/${mcVersion}/builds/${buildNum}/downloads/paper-${mcVersion}-${buildNum}.jar";
```
Would it be possible to change `version` to look like `1.20.2-234` instead of `1.20.2.234`?

That would eliminate the parsing logic, and I think it could be a easier for users (who may be confused by the extra decimal point)

MayNiklas replied:
That sounds like a great idea!
My goal was, to firstly migrating the package 1:1 before changing it's logic.
One disadvantage I can think of: we would break setups, using some very specific overrides.

I will add it to my PR.

MayNiklas replied:
Edit: why does it eliminate the parsing logic?
We still have to use build number and version separately in the URL.

MayNiklas replied:
I implemented the suggested changes!
If you like it this way, I would squash the changes to a single commit.

aaronjanse replied:
Oops, I didn't see that the version still needs to be split.

Do we think a lot of users are using specific overrides? I don't want to break users' setups if that's the case

MayNiklas replied:
Puh, hard to say...
Generally speaking: pinning a specific version is quite important when hosting a Minecraft server.

That being said:
I don't think that is a big problem.
Because of the nix store (hash does not change), only new installs would fail. People using overrides would easily be able to look up what has changed.
As I demonstrated above, overrides get simplified by this PR too!

In my opinion:
This syntax makes it easier for new users.
People already having overrides will know how to change them, especially since they are now easier... (we could even add a comment with an example in theory?).

aaronjanse reviewed (approved):
Thank you so much for the patience @MayNiklas.

Everything here looks good to me!

MayNiklas commented:
> Thank you so much for the patience @MayNiklas.
> 
> Everything here looks good to me!

Great!
I rebased both commits into a single commit to simplify merging once we are ready.

MayNiklas commented:
@aaronjanse Right now, I'm asking myself:
would it make sense to change this PR to a approach similar to this one: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/games/minecraft-servers/versions.json

I played around with it locally and it works.
Big advantage: having the versions in a json file, makes it easier to update the package with update scripts 🤔
What do you think?
Should I change this?

MayNiklas commented:
@aaronjanse I added my suggestion to this PR.
Tell me, if you think it makes sense (I would stash my branch).
Otherwise, I will revert the last commit.

It would make this package more similar to the MC package and allow easier updates.

aaronjanse reviewed (approved):
This looks good to me!

Thank you @MayNiklas for the change, and I'm glad it'll be easier to update versions going forward 🎉

Scrumplex commented:
Result of `nixpkgs-review pr 276292` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>papermc (papermcServers.papermc ,papermcServers.papermc-1_20_4)</li>
    <li>papermcServers.papermc-1_18_2</li>
    <li>papermcServers.papermc-1_19_4</li>
    <li>papermcServers.papermc-1_20_2</li>
  </ul>
</details>

MayNiklas commented:
I also have a update script ready that I will add once this PR got merged - it will automatically create the versions.json file with versions newer than 1.18.

I decided against adding it to this PR since I want to avoid modifying this PR so it can get merged faster.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -38174,7 +38174,9 @@ with pkgs;
   pacvim = callPackage ../games/pacvim { };
 
-  papermc = callPackage ../games/papermc { };
+  papermcServers = import ../games/papermc { inherit callPackage lib; };
```
```suggestion
  papermcServers = callPackages ../games/papermc { };
```

SuperSandro2000 commented on pkgs/games/papermc/derivation.nix:
```diff
@@ -0,0 +1,45 @@
+{ lib, stdenvNoCC, fetchurl, makeBinaryWrapper, jre, version, hash, ... }:
```
```suggestion
{ lib, stdenvNoCC, fetchurl, makeBinaryWrapper, jre, version, hash }:

```

MayNiklas committed:
papermc: provide multiple versions

MayNiklas commented:
I did the following:
* rebased on master
* implemented the changes requested by @SuperSandro2000
* added a `update.py` script that uses `https://api.papermc.io/v2/projects/paper` to get information about new versions
